Lib/importlib/_bootstrap.py

```diff
@@ -1016,31 +1016,30 @@ def _handle_fromlist(module, fromlist, import_, *, recursive=False):
     """
     # The hell that is fromlist ...
     # If a package was imported, try to import stuff from fromlist.
-    if hasattr(module, '__path__'):
-        for x in fromlist:
-            if not isinstance(x, str):
-                if recursive:
-                    where = module.__name__ + '.__all__'
-                else:
-                    where = "``from list''"
-                raise TypeError(f"Item in {where} must be str, "
-                                f"not {type(x).__name__}")
-            elif x == '*':
-                if not recursive and hasattr(module, '__all__'):
-                    _handle_fromlist(module, module.__all__, import_,
-                                     recursive=True)
-            elif not hasattr(module, x):
-                from_name = '{}.{}'.format(module.__name__, x)
-                try:
-                    _call_with_frames_removed(import_, from_name)
-                except ModuleNotFoundError as exc:
-                    # Backwards-compatibility dictates we ignore failed
-                    # imports triggered by fromlist for modules that don't
-                    # exist.
-                    if (exc.name == from_name and
-                        sys.modules.get(from_name, _NEEDS_LOADING) is not None):
-                        continue
-                    raise
+    for x in fromlist:
+        if not isinstance(x, str):
+            if recursive:
+                where = module.__name__ + '.__all__'
+            else:
+                where = "``from list''"
+            raise TypeError(f"Item in {where} must be str, "
+                            f"not {type(x).__name__}")
+        elif x == '*':
+            if not recursive and hasattr(module, '__all__'):
+                _handle_fromlist(module, module.__all__, import_,
+                                 recursive=True)
+        elif not hasattr(module, x):
+            from_name = '{}.{}'.format(module.__name__, x)
```

Comment:
The attempt to import missing names as submodules should still be restricted to modules with a `__path__` attribute. Something like `try_import = hasattr(module, "__path__")` before the loop, and then modify the condition on this branch to `elif try_import and not hasattr(module, x):`

Reply:
(Since we know at the point of calling whether `module` is a package or not, `try_import` could also be a parameter that defaults to `False`, with `__import__` setting it to `True` as part of the call)

Reply:
The point of this issue is that this check is moved out from this function to calling places.

Reply:
In the issue for improving the error message for non-str objects in regular modules, calling this is one of the options for handling the enhancement. However, we only want to allow the first two branches for that use case, not the nested import one.

Reply:
Yes, I suggested this. But it would prevent the optimization made by this PR.